### PR TITLE
Selam Ainalem Amperes

### PR DESCRIPTION
**Pull request by SelamawitA**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I heavily relied on postman to play with the API as well as reading the documentation. On parts of the documentation that weren't as clear (r = uri for example) copying the examples and running the request myself in Postman made the connections and the requirements more clear. 
Describe your API Wrapper.  How did you decide on the methods you created? | My design was driven by relying on the methods provided by the API, realistically I know that more API calls = $$ but it seemed more efficient for this project to rely on the API for calling a single recipe, list from a query, range of queries over pulling on large set of a search, storing it somewhere and having to constantly search. 
Describe an edge case or failure case test you wrote for your API Wrapper. | The case of a non-string type entered for a query. I added a condition to return an empty array in the event that this occurs. 
Explain how VCR aids in testing an API. | VCR's store calls to APIs made by tests so if tests need to be run/updated multiple times the data from the original API call can be referenced, limiting the overall cost of API usage.  
What is the Heroku URL of your deployed application? | https://limitless-wildwood-28723.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/N2UFecmf/recipeapi

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+
+    <title><%= content_for?(:title) ? yield(:title) : "Untitled" %></title>
+
+    <%= stylesheet_link_tag  "application" %>
+    <%= javascript_include_tag "application", 'data-turbolinks-track' => true %>
+    <%= csrf_meta_tags %>
+  </head>
+  <nav>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:13]: This `nav` element should be in the `body`

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  <%=link_to "Home", root_path ,class:"button"%>
+<%if user = User.find_by(id: session[:user_id]) %>
+  <span id="user-signed-in">Signed in as: <span id="user-name"><%=user.username%></span> </span><%=link_to "Log out", signout_path, class:"button"%>
+<%else%>
+  <%= link_to "Sign-in with Github", "/auth/github", class:"button"  %>
+<%end%>
+  </nav>
+
+  <%if params[:query].nil?%>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:22]: Maybe simply put `main` elements with these ids in the views for homepage and search etc.

File: app/views/recipe/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+
+<h2 class="item-header"><%=@recipe.label%></h2>
+<h3 class="item-header">By <%=@recipe.source%></h3>
```
Comment (CheezItMan) [inline app/views/recipe/show.html.erb:3]: This would be an excellent point for a link back to the source of the recipe.

File: app/controllers/recipe_controller.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      else
+        flash[:status] = :success
+        flash[:message] = "Sucecssfully found #{@query} recipes."
+      end
+    end
+
+  end
+
+  def show
+    @recipe = RecipeApiWrapper.a_recipe(params[:uri])
```
Comment (CheezItMan) [inline app/controllers/recipe_controller.rb:34]: You should have some error handling for if a recipe isn't found.

File: lib/recipe_api_wrapper.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    return list_of_recipes
+  end
+
+  def self.a_recipe(recipe)
+    if recipe.class != String
+      return []
+    end
+    a_recipe = recipe.uri_encode
+    response = HTTParty.get("https://api.edamam.com/search?app_id=#{APP_ID}&app_key=#{APP_KEY}&r=#{a_recipe}")
```
Comment (CheezItMan) [inline lib/recipe_api_wrapper.rb:28]: You should do some error handling for if the API request doesn't work.  What if the URI is incorrect?  How do you handle it?

File: test/lib/recipe_api_wrapper_test.rb [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+          #recieve expected object type
+          response.class.must_equal Recipe
+
+          #recieve expected response
+          response.label.must_equal "Pizza"
+          response.source.must_equal "Cooking Channel"
+        end
+      end
+      it 'will return [] for a recipe search that is not of string type' do
+        VCR.use_cassette("recipe-is-a-string") do
```
Comment (CheezItMan) [inline test/lib/recipe_api_wrapper_test.rb:55]: What about a call with a URI that's not valid?

File: lib/recipe_api_wrapper.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    recipe = Recipe.new(response[0])
+    return recipe
+  end
+
+  def self.limited_query(query,from,to)
```
Comment (CheezItMan) [inline lib/recipe_api_wrapper.rb:34]: This looks like a great candidate for combining with the `list_of_queried_recipes` method using default parameters for the from and to fields.

File: test/controllers/recipe_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          must_respond_with :ok
+          flash[:status].must_equal :success
+          flash[:message].must_equal "Sucecssfully found chicken recipes."
+        end
+      end
+
+      it 'will display index for query within a specific range' do
+        session[:user_id].must_equal @jay.id
+        VCR.use_cassette("Recipe-controller:recipe-query-range") do
+          get query_results_path, params: {from:"1", to: "5", query: "corn"}
```
Comment (CheezItMan) [inline test/controllers/recipe_controller_test.rb:33]: You should also test with invalid `from` and `to` fields

File: test/controllers/recipe_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        session[:user_id].must_equal @jay.id
+        VCR.use_cassette("recipe-query-no-result") do
+          get query_results_path, params: {query: "jjjjj"}
+          flash[:status].must_equal :error
+          flash[:message].must_equal "We currently do not have any jjjjj recipes."
+        end
+      end
+    end
+    describe 'Show' do
+      it 'will show details of a recipe' do
```
Comment (CheezItMan) [inline test/controllers/recipe_controller_test.rb:48]: What about invalid `uri`s

File: test/controllers/recipe_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,84 @@
+require "test_helper"
+
+describe RecipeController do
+  before do
+    @jay = User.create(username:"jay-z",email:"jayz@gmail.com", uid: 132123)
+  end
+  describe 'Logged-in User' do
+    before do
+      login(@jay)
```
Comment (CheezItMan) [inline test/controllers/recipe_controller_test.rb:9]: Very good that you're doing this testing.

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | good number of commits and good commit messages
Comprehension questions | Check
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Check
Semantic HTML | You've got some broken HTML in your layout file
Errors are reported to the user | For search results with no recipes, yes.  For incorrect URIs for the show action, no.  
API Wrapper to handle the API requests | Check, see my notes in your code
Controller testing | Check, 
Lib testing | Check, see my notes
Search Functionality | Check
List Functionality | Check
Show individual item functionality (link to original recipe opens in new tab) | Check
**Styling** | 
Responsive layout | The search results are nicely responsive, the show actions, not so much
List View shows 10 items at a time/pagination | Check
The app is styled to create an attractive user interface | The styling looks pretty good.
**API Features** | 
The App attributes Edaman | Check
The VCR casettes do not contain the API key | Check, nice that you're also filtering the github ID and secret.
**External Resources** | 
Link to Trello Board | Yes, but I can't access it.
Link to deployed app on Heroku | Check
**Overall** | Nice work, you hit the major learning goals.  You have some things to work on including through testing and making sure you handle errors, but you did overall a nice job here.
